### DISCUSSION TITLE
How to Crop with CLI?

### DISCUSSION THREAD

**Discussion by ivanjaros** [libvips/libvips#4257] [Q&A] [answered]
I want to crop image while specifying location of the crop. Like center, top, left, top-left and so on. The smartcrop seems to be useless and the crop requires me to know all the dimensions, which i don't. Is there some other cropping method I can use to just do "crop image.jpg into 500x500 with focus on top-right corner"?

Using smartcrop with `--attention-x 100000 --attention-y 100000` for top right corner, for example, is not working and I always get center crop. If i change `interesting` to something else, i always get too many arguments error, so the coordinates and method are not cooperating.

**Comment by jcupitt** [comment]
Hi again,

You can use `vipsheader` to read fields from image metadata, eg.:

```
$ vipsheader -f width ~/pics/k2.jpg
1450
```

So that jpg is 1450 pixels across. Use `vipsheader -a k2.jpg` to see all fields. 

In bash you could write:

```bash
width=$(vipsheader -f width $1)
vips crop $1 $2 $(($width - 500)) 0 500 500
```

**Comment by ivanjaros** [comment]
I wanted to avoid having to read the file in order to get the dimensions and do the math manually, but if there is no other way...it is what it is.

**Comment by jcupitt** [comment]
Smartcrop ought to work. For example, with this image:

![owl](https://github.com/user-attachments/assets/7a263a3f-fc49-4976-9861-c0b35277db0f)

I can run:

```
vips smartcrop owl.jpg x.png 300 300
```

To make:

![x](https://github.com/user-attachments/assets/e9a32739-bf61-4e65-8af0-a8d2c406d223)

And it's kind-of found the owl's head.

It's pretty crude and won't always work, of course.

**Comment by jcupitt** [comment, marked as answer]
> I wanted to avoid having to read the file in order to get the dimensions

`vipsheader` is pretty quick, it only reads the image header and the image metadata.

**Comment by jcupitt** [comment]
> --attention-x 100000 --attention-y 100000 

Ah, those arguments are for output. You can get some help with eg.:

```
$ vips smartcrop
extract an area from an image
usage:
   smartcrop input out width height [--option-name option-value ...]
where:
   input        - Input image, input VipsImage
   out          - Output image, output VipsImage
   width        - Width of extract area, input gint
			default: 1
			min: 1, max: 100000000
   height       - Height of extract area, input gint
			default: 1
			min: 1, max: 100000000
optional arguments:
   attention-x  - Horizontal position of attention centre, output gint
			default: 0
			min: 0, max: 100000000
   attention-y  - Vertical position of attention centre, output gint
			default: 0
			min: 0, max: 100000000
   interesting  - How to measure interestingness, input VipsInteresting
			default enum: attention
			allowed enums: none, centre, entropy, attention, low, high, all
   premultiplied - Input image already has premultiplied alpha, input gboolean
			default: false
```

They are both "output gint". You use them like this:

```
$ vips smartcrop owl.jpg x.png 300 300 --attention-x --attention-y
337
111
```

They make the operation output the coordinates of the crop box it selected.

**Comment by ivanjaros** [comment]
I am essentially trying to target crop like this:
![obrázok](https://github.com/user-attachments/assets/c0fb4fbf-5a6f-4c85-93d4-9e7b34434739)

So I have some source Image and I want to take a crop of xWidth and yHeight and from a region in the source image(tl, t, tr, l, c, r, bl, b, br).
This is because i am trying to keep compatibility with previous code, which was native, and now i am moving to vips via cli.

**Reply by ivanjaros** [reply]
i aready implemented the code, where i use vipsheader, get dimensions, and generate coordinates for vips_crop. just fyi.

**Reply by jcupitt** [reply]
Is this from Go? There are libvips bindings for Go (though I've not used them myself), they might be easier to use, faster, and need less memory than the CLI. 

The libvips CLI is deliberately very basic. I don't much like the imagemagick-style CLI interface where it's so complex it's almost its own programming language. The idea for libvips was to just do image processing, and to make it very simple to write bindings. That way, people could use their existing programming languages for the logic and not have to learn a whole new thing just for images. Plus I'd avoid having to maintain a lot of code.

libvips has full introspection, so bindings just need to wrap (I think?) 20 API calls, and they get all 300 or so libvips operations. Plus since the binding is effectively generated at runtime from introspection of the library binary, it'll automatically update as libvips updates and should need almost no maintenance. The core of pyvips (for example) is under 300 lines of pure python code.

**Reply by ivanjaros** [reply]
yes, i am using Go but i don't allow cgo in my codebase. so i am using the cli version.

**Reply by jcupitt** [reply]
You could shell out to a python program that implemented the operations you need. It'd be a bit ugly, but faster and easier than the CLI.

**Reply by ivanjaros** [reply]
why would it be faster? cli is calling the vips directly. calling python to call vips c api is just adding unnecessary layer.

**Reply by jcupitt** [reply]
We'd need to make a benchmark that represents your typical workflow and time it, but if you are doing a series of operations, python is a lot quicker since it can chain them together. You get much more parallelism, and you save all the IO.

**Reply by jcupitt** [reply]
... there's a benchmark here, for example:

https://github.com/libvips/libvips/wiki/Speed-and-memory-use

The test is: load a 10k x 10k pixel TIFF, crop 100 pixels off every edge, shrink by 10% with bilinear interpolation, sharpen with a 3x3 convolution and save again. pyvips is about twice as fast and needs  one third of the IO. You'd see larger gains with a more complex chain of operations.
